// firebase-nightlight/dist/mock-database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var mock_error_1 = require("./mock-error");
var mock_ref_1 = require("./mock-ref");
var MockDatabase = (function () {
    function MockDatabase(options) {
        this.app_ = options.app;
        this.database_ = options.database || {};
        this.database_.content = this.database_.content || {};
        this.emitters_ = options.emitters;
    }
    Object.defineProperty(MockDatabase.prototype, "app", {
        get: function () {
            return this.app_;
        },
        enumerable: true,
        configurable: true
    });
    MockDatabase.prototype.goOffline = function () {
        throw mock_error_1.unsupported_();
    };
    MockDatabase.prototype.goOnline = function () {
        throw mock_error_1.unsupported_();
    };
    MockDatabase.prototype.ref = function (path) {
        return new mock_ref_1.MockRef({
            app: this.app_,
            database: this.database_,
            emitters: this.emitters_,
            path: path || null
        });
    };
    MockDatabase.prototype.refFromURL = function (url) {
        return new mock_ref_1.MockRef({
            app: this.app_,
            database: this.database_,
            emitters: this.emitters_,
            path: url.replace(/^(http|https):\/\/[^\/]+/i, "")
        });
    };
    return MockDatabase;
}());
exports.MockDatabase = MockDatabase;
